tests/TestCase/Database/CommonTableExpressionIntegrationTest.php

```diff
@@ -0,0 +1,942 @@
+            $query->sql(),
+            !$this->autoQuote
+        );
+
+        $expected = [
+            [
+                'col1' => '1',
+                'col2' => '2',
+            ],
+        ];
+
+        $stmt = $query->execute();
+        $result = $stmt->fetchAll('assoc');
+        $stmt->closeCursor();
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testWithFieldAsExpressions(): void
+    {
+        $cteQuery = $this->connection
+            ->newQuery()
+            ->select([1, 2]);
+
+        $query = $this->connection
+            ->newQuery()
+            ->with(function (QueryExpression $exp, Query $query) use ($cteQuery) {
+                return $query
+                    ->cte('cte', $cteQuery)
+                    ->setFields([
+                        $cteQuery->identifier('col1'),
+                        $cteQuery->identifier('col2'),
+                    ]);
+            })
+            ->select(['col1', 'col2'])
+            ->from('cte');
+
+        $this->assertQuotedQuery(
+            'WITH cte\(<col1>, <col2>\) AS \(SELECT 1, 2\) SELECT <col1>, <col2> FROM <cte>',
+            $query->sql(),
+            !$this->autoQuote
+        );
+
+        $expected = [
+            [
+                'col1' => '1',
+                'col2' => '2',
+            ],
+        ];
+
+        $stmt = $query->execute();
+        $result = $stmt->fetchAll('assoc');
+        $stmt->closeCursor();
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testWithReset(): void
+    {
+        $query = $this->connection->newQuery();
+
+        $this->assertEmpty($query->clause('with'));
+
+        $query
+            ->with(function (QueryExpression $exp, Query $query) {
+                return $query
+                    ->cte('cte1', $query->getConnection()->newQuery()->select(['col' => 1]));
+            })
+            ->select('col')
+            ->from('cte');
+
+        $this->assertCount(1, $query->clause('with'));
+        $this->assertQuotedQuery(
+            'WITH cte1 AS \(SELECT 1 AS <col>\) SELECT <col> FROM <cte>',
+            $query->sql(),
+            !$this->autoQuote
+        );
+
+        $query->with(null, true);
+
+        $this->assertEmpty($query->clause('with'));
+        $this->assertQuotedQuery(
+            'SELECT <col> FROM <cte>',
+            $query->sql(),
+            !$this->autoQuote
+        );
+    }
+
+    public function testWithResetRequiresOverwriting(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Resetting the WITH clause only works when overwriting is enabled.');
+
+        $this->connection->newQuery()->with(null);
+    }
+
+    public function testWithOverwrite(): void
+    {
+        $query = $this->connection->newQuery();
+
+        $this->assertEmpty($query->clause('with'));
+
+        $query
+            ->with(function (QueryExpression $exp, Query $query) {
```

Comment:
If the CommonTableExpression had setName() and setQuery() methods would we need Query::cte()?

Reply:
I wonder if we should follow the sql builder convention of `query()` and `modifier()`.

Reply:
Someone should probably write down the required conventions, AFAIR it was to use non `get/set` in the query builder only? Guess it's debatable whether expressions are close enough to the builder to be affected by that convention, but currently it seems that expressions generally use the `get/set` notation.

Reply:
We generally don't use get/set in the query builder, as we didn't want to add 'noise words' when you're dealing with SQL concepts. The builder is generally all 'set' anyways.

Reply:
I take that as a "*use get/set for expressions*"...